### PR TITLE
[UDF] JSON Array Contains function

### LINKED ISSUE #607: UDF function to search if the value exists in the json array
Currently, JSON UDF functionality is limited to only EXTRACTJSONFIELD function. However, it will be great to allow users functionality to probe a value inside JSON array.

Example:
```
SELECT JSON_ARRAY_CONTAINS('[1,2,3]', 3) FROM DUAL
true
```

### PR DESCRIPTION
**Pull request by satybald**
UDF function that searches if the value exists in JSON Array. Function accepts two arguments a JSON string, and a searchValue. SearchValue can be any of the  BOOLEAN, INTEGER, BIGINT, DOUBLE and VARCHAR. Fixes #607

### CODE REVIEW

**Comment by bluemonk3y** [comment]
@satybald - there is an internal nexus issue - it will be resolved in the next couple of hours

File: ksql-engine/src/main/java/io/confluent/ksql/function/FunctionRegistry.java [hunk 1]
```diff
@@ -151,6 +153,12 @@ private void init() {
         "EXTRACTJSONFIELD", JsonExtractStringKudf.class);
     addFunction(getStringFromJson);
 
+    Schema array = SchemaBuilder.array(Schema.STRING_SCHEMA).build();
+    KsqlFunction jsonArrayContainsString = new KsqlFunction(
+            Schema.BOOLEAN_SCHEMA, Arrays.asList(array, Schema.STRING_SCHEMA),
+            "JSON_ARRAY_CONTAINS", JsonArrayContainsKudf.class);
+    addFunction(jsonArrayContainsString);
```
Comment (satybald) [inline ksql-engine/src/main/java/io/confluent/ksql/function/FunctionRegistry.java]: I have an issue registering my function passing arguments to it. I'm expecting that 'evaluate' method passes json string as first argment of the array, and the second argument is search value.

However, when I execute fullowing query: `select JSON_ARRAY_CONAINS(["abc", "aaa", "dd"], 'aaa') from whatever`.  evaluate function gets the string array of ["abc", "aaa", "dd"].

**Comment by satybald** [comment]
thanks @bluemonk3y for information. Is there any way of getting the result of Jenkins build?

**Comment by hjafarpour** [comment]
@satybald I would suggest to change this function to search general array type regardless if the serialization mode is JSON or AVRO. This way this can be used in more broader use cases.

**Comment by satybald** [comment]
@hjafarpour thanks for the suggestion. I added support for the general array type.

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/JsonArrayContainsKudfTest.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    private JsonArrayContainsKudf jsonUdf = new JsonArrayContainsKudf();
+
+    @Override
+    public void setUp()
+            throws Exception
+    {
+        jsonUdf.init();
+    }
+
+    public void testEmptyArray()
```
Comment (hjafarpour) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/JsonArrayContainsKudfTest.java]: '@ Test` annotation?

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/JsonArrayContainsKudfTest.java [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+        assertEquals(true, jsonUdf.evaluate(json, -2147483648));
+        assertEquals(true, jsonUdf.evaluate(json, 1));
+        assertEquals(true, jsonUdf.evaluate(json, 2));
+        assertEquals(true, jsonUdf.evaluate(json, 3));
+        assertEquals(false, jsonUdf.evaluate("5", 5));
+        assertEquals(false, jsonUdf.evaluate(json, 5));
+    }
+
+    public void testIfJsonArrayContainsLongValue()
+    {
```
Comment (hjafarpour) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/JsonArrayContainsKudfTest.java]: As above, missing Test annotation.

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/JsonArrayContainsKudfTest.java [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+        assertEquals(true, jsonUdf.evaluate("[1111111111111111]", 1111111111111111L));
+        assertEquals(true, jsonUdf.evaluate("[[222222222222222], 33333]", 33333L));
+        assertEquals(true, jsonUdf.evaluate("[{}, \"abc\", null, 1]", 1L));
+        assertEquals(false, jsonUdf.evaluate("[[222222222222222], 33333]", 222222222222222L));
+        assertEquals(false, jsonUdf.evaluate("[{}, \"abc\", null, [1]]", 1L));
+        assertEquals(false, jsonUdf.evaluate("[{}, \"abc\", null, {\"1\":1}]", 1L));
+    }
+
+    public void testIfArrayContainsDoubleValue()
+    {
```
Comment (hjafarpour) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/JsonArrayContainsKudfTest.java]: As above, missing Test annotation.

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/JsonArrayContainsKudfTest.java [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+        assertEquals(true, jsonUdf.evaluate("[-1.0, 2.0, 3.0]", 2.0));
+        assertEquals(true, jsonUdf.evaluate("[1.0, -2.0, 3.0]", -2.0));
+        assertEquals(true, jsonUdf.evaluate("[1.0, 2.0, 1.6E3]", 1600.0));
+        assertEquals(true, jsonUdf.evaluate("[1.0, 2.0, -1.6E3]", -1600.0));
+        assertEquals(true, jsonUdf.evaluate("[{}, \"abc\", null, -2.0]", -2.0));
+        assertEquals(false, jsonUdf.evaluate("[[2.0], 3.0]", 2.0));
+    }
+
+    public void testIfArrayContainsStringValue()
```
Comment (hjafarpour) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/JsonArrayContainsKudfTest.java]: As above, missing Test annotation.
The same for all tests methods.

**Comment by satybald** [comment]
@hjafarpour thanks for the feedback. Added @Test annotation.

File: ksql-engine/src/main/java/io/confluent/ksql/function/FunctionRegistry.java [hunk 6]
```diff
@@ -151,6 +153,27 @@ private void init() {
         "EXTRACTJSONFIELD", JsonExtractStringKudf.class);
     addFunction(getStringFromJson);
 
+    KsqlFunction jsonArrayContainsString = new KsqlFunction(
+            Schema.BOOLEAN_SCHEMA, Arrays.asList(Schema.STRING_SCHEMA, Schema.STRING_SCHEMA),
+            "JSON_ARRAY_CONTAINS", JsonArrayContainsKudf.class);
```
Comment (satybald) [inline ksql-engine/src/main/java/io/confluent/ksql/function/FunctionRegistry.java]: I have doubts should name contains underscore or not? For consistency with `EXTRACTJSONFIELD` it should be without, but for readability, I argue that underscore is needed.

Comment (bluemonk3y) [reply ksql-engine/src/main/java/io/confluent/ksql/function/FunctionRegistry.java]: @satybald  - it needs to be consistent for this PR. EXTRACTJSONFIELD - changing formatting will require doc changes etc - which could be done in another PR

Comment (satybald) [reply ksql-engine/src/main/java/io/confluent/ksql/function/FunctionRegistry.java]: @bluemonk3y good point about the doc. I completely forgot to document function in the doc :smile:

**Comment by satybald** [comment]
@hjafarpour @dguy what do you think about PR? I compile and test branch locally build passes all the tests.

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/json/ArrayContainsKudf.java [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+  }
+
+  @Override
+  public Object evaluate(Object... args) {
+    if (args.length != 2) {
+      throw new KsqlFunctionException("ARRAY_CONTAINS udf should have two input argument. Given: " + Arrays.toString(args));
+    }
+    Object searchValue = args[1];
+    if(args[0] instanceof String) {
+      String jsonString = args[0].toString();
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/json/ArrayContainsKudf.java]: perhaps just cast it to string and inline in `ifJsonStringArrayContains`

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/json/ArrayContainsKudf.java [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+      return ifJsonStringArrayContains(jsonString, searchValue);
+    } else if(args[0] instanceof Object[]) {
+      Object[] array = (Object[]) args[0];
+      return ifArrayContains(array, searchValue);
+    }
+    throw new KsqlFunctionException("Invalid type parameters for " + Arrays.toString(args));
+  }
+
+  private Object ifArrayContains(Object[] array, Object searchValue) {
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/json/ArrayContainsKudf.java]: return `boolean` and rename to `arrayContains`

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/json/ArrayContainsKudf.java [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    //TODO: Refactor this to ArrayUtil.containsValue from TopK PR#575
+    for (Object value : array) {
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/json/ArrayContainsKudf.java]: `return Arrays.stream(array).anyMatch(value -> Objects.equals(o, searchValue))`

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/json/ArrayContainsKudf.java [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+      if(Objects.equals(value, searchValue)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean ifJsonStringArrayContains(String json, Object searchValue) {
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/json/ArrayContainsKudf.java]: rename to `stringArrayContains`

File: ksql-engine/src/main/java/io/confluent/ksql/function/udf/json/ArrayContainsKudf.java [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    JsonToken valueType = getType(searchValue);
+    try (JsonParser parser = JSON_FACTORY.createParser(json)) {
+      if (parser.nextToken() != START_ARRAY) {
+        return false;
+      }
+
+      while (parser.currentToken() != null) {
+        JsonToken token = parser.nextToken();
+        if (token == null) {
+          return token == searchValue;
```
Comment (dguy) [inline ksql-engine/src/main/java/io/confluent/ksql/function/udf/json/ArrayContainsKudf.java]: `return searchValue == null` ?

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/ArrayContainsKudfTest.java [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+import org.junit.Test;
+
+public class ArrayContainsKudfTest
+        extends TestCase
+{
+    private ArrayContainsKudf jsonUdf = new ArrayContainsKudf();
+
+    @Override
+    public void setUp()
+            throws Exception
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/ArrayContainsKudfTest.java]: you don't need the `throws Exception`

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/ArrayContainsKudfTest.java [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+    {
+        jsonUdf.init();
+    }
+
+    @Test
+    public void testEmptyArray()
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/ArrayContainsKudfTest.java]: i prefer test names that tell me what it is the object should do, i.e, `shouldReturnFalseOnEmptyArray`

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/ArrayContainsKudfTest.java [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+    {
+        assertEquals(false, jsonUdf.evaluate("[]", true));
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/ArrayContainsKudfTest.java]: `assertFalse` for all of these

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/ArrayContainsKudfTest.java [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+        assertEquals(false, jsonUdf.evaluate("[]", false));
+        assertEquals(false, jsonUdf.evaluate("[]", null));
+        assertEquals(false, jsonUdf.evaluate("[]", 1.0));
+        assertEquals(false, jsonUdf.evaluate("[]", 100));
+        assertEquals(false, jsonUdf.evaluate("[]", "abc"));
+        assertEquals(false, jsonUdf.evaluate("[]", ""));
+    }
+
+    @Test
+    public void testNullArray()
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/ArrayContainsKudfTest.java]: split this into two tests, i.e., one for when it should be true and on for when it should be false. Also naming as proposed above

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/ArrayContainsKudfTest.java [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+    {
+        assertEquals(true, jsonUdf.evaluate("[null]", null));
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/ArrayContainsKudfTest.java]: `assertTrue(...)` etc

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/ArrayContainsKudfTest.java [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+        assertEquals(false, jsonUdf.evaluate("[null]", true));
+        assertEquals(false, jsonUdf.evaluate("[null]", false));
+        assertEquals(false, jsonUdf.evaluate("[null]", 1.0));
+        assertEquals(false, jsonUdf.evaluate("[null]", 100));
+        assertEquals(false, jsonUdf.evaluate("[null]", "abc"));
+        assertEquals(false, jsonUdf.evaluate("[null]", ""));
+    }
+
+    @Test
+    public void testIntegerJsonArrayContainsValues()
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/ArrayContainsKudfTest.java]: same as above

File: ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/ArrayContainsKudfTest.java [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+        assertEquals(true, jsonUdf.evaluate(json, -2147483648));
+        assertEquals(true, jsonUdf.evaluate(json, 1));
+        assertEquals(true, jsonUdf.evaluate(json, 2));
+        assertEquals(true, jsonUdf.evaluate(json, 3));
+        assertEquals(false, jsonUdf.evaluate("5", 5));
+        assertEquals(false, jsonUdf.evaluate(json, 5));
+    }
+
+    @Test
+    public void testIfJsonArrayContainsLongValue()
```
Comment (dguy) [inline ksql-engine/src/test/java/io/confluent/ksql/function/udf/json/ArrayContainsKudfTest.java]: same as above

**Comment by satybald** [comment]
@dguy thanks for the review. I've applied your feedback. However, I cannot use assertTrue() and assertFalse() as jsonUdf.evaluate returns Object. I need explicitly state against which object I'm evaluating.

**Comment by dguy** [comment]
@satybald with your latest updates it looks like you have bought in a bunch of changes that aren't yours, i.e, Dockerfile, pom.xml etc

**Comment by satybald** [comment]
@dguy I did rebase with a master to add `ArraysUtill.containsValue`. However, I forgot that there's a lot of changes since then has been merged to master.

**Comment by dguy** [comment]
@satybald - those changes shouldn't show up in this diff, though. They should already be in master

**Comment by satybald** [comment]
retest this please

**Comment by satybald** [comment]
@dguy I fixed that now and cherry-picked need me commit with `ArraysUtill.containsValue` changes. Could you please look once more?

**Comment by bluemonk3y** [comment]
@satybald  ..engine/src/main/java/io/confluent/ksql/function/udf/json/ArrayContainsKudf.java:29:8: Unused import - java.util.Objects. [UnusedImports]

**Comment by satybald** [comment]
thanks, @bluemonk3y. fixed the issue.

**Comment by satybald** [comment]
@dguy @bluemonk3y @hjafarpour build again failed :cry:

**Comment by dguy** [comment]
Thanks @satybald, the build has passed so i've merged this PR. Thanks for the contribution

**Pull request merged**